File: Day01/part2.py
def main():
    with open('./input.txt') as file:
        left_nums = []
        right_counts = {}

        for line in file:
            line = line.strip()
            tokens = line.split()
            numbers = [int(token) for token in tokens]
            left, right = numbers

            left_nums.append(left)

            if right not in right_counts:
                right_counts[right] = 1
            else:
                right_counts[right] = right_counts[right] + 1

    sim_score = 0

    for num in left_nums:
        count = 0

        if num in right_counts:
            count = right_counts[num]

        sim_score += (num * count)

    print(sim_score)

if __name__ == '__main__':
    main()

File: Day01/part2.pl
use strict;
use warnings;
use List::Util 'sum';

my @left_nums;
my %right_counts;

open (my $file, "<", "input.txt") or die $!;

while (my $line = <$file>) {
    my @nums = split(" ", $line);
    my $left = $nums[0];
    my $right = $nums[1];

    unless (exists $right_counts{$right}) {
        $right_counts{$right} = 0
    }

    push(@left_nums, $left);
    $right_counts{$right}++;
}

close ($file);

my @sim_scores = map {exists $right_counts{$_} ? $_ * $right_counts{$_} : 0} @left_nums;
my $sim_score = sum(@sim_scores);

print("Total: $sim_score\n");
